flocker/node/agents/functional/test_cinder_behaviour.py

```diff
@@ -112,3 +127,39 @@ def test_updated_metadata_is_listed(self):
         missing_items = expected_items - actual_items
 
         self.assertEqual(set(), missing_items)
+
+
+class BlockDeviceAPIDestroyTests(SynchronousTestCase):
+    """
+    Test for ``cinder.CinderBlockDeviceAPI.destroy_volume``
+    """
+    def setUp(self):
+        self.cinder_volumes = cinder_volume_manager()
+
+    def test_destroy_timesout(self):
+        """
+        If the cinder cannot delete the volume, we should timeout
+        after waiting some time
+        """
+        # Adding a fake get that will always return a volume, so we
+        # timeout waiting for the volume to be deleted
+        self.patch(self.cinder_volumes, "get", fakeGet)
+        # Using a fake no-op delete so it doens't actually delete anything
+        # (we don't need any actual volumes here, as we only need to verify
+        # the timeout)
+        self.patch(self.cinder_volumes, "delete", lambda *args, **kwargs: None)
+
+        api = CinderBlockDeviceAPI(
+            cinder_volume_manager=self.cinder_volumes,
+            nova_volume_manager=object(),
+            nova_server_manager=object(),
+            cluster_id=uuid4()
+            )
+        # Setting the timeut to 1, as the default is quite high, and we do not
+        # want to wait that much in a test
+        api.timeout = 1
+        self.assertRaises(
+            TimeoutException,
+            api.destroy_volume,
+            blockdevice_id=unicode(uuid4())
+        )
```

Comment:
This demonstrates a timeout happens but doesn't demonstrate anything about when it happens.  The specified timeout value might be ignored and the timeout raised after 0.1 seconds or 1000 seconds... The test passes either way.

It would be nice to make some assertion about time here.  This is another place where a fake clock is nice since it's rather difficult to make any reliable assertions about the result of `time.time()`.  Even if you assert something like "the time after the exception is no later than 10 seconds after the time before the call was made" you still run the risk that someone will run the test on a loaded machine where 10 seconds actually do elapse in between those events even though the timeout of 1 second is being properly respected by the implementation.

One option (not something we've tried in flocker before afaik) might be to add some logging and try making the time assertions based on the contents of the log messages.  For example, if the loop logged the Eliot equivalent to "giving up because at current time=x is after planned timeout time=y, deletion not achieved within timeout=z seconds of starting time=a" and then you can check the logged values and see if the logging logic at least reflects the right timeout behavior.

Another option would be to try really using a fake clock.  Maybe you can avoid needing threads if you use something like a fake that reports real time but records the values it reports back to a caller.  Then the timeout logic is satisfied if the last reported time is the first time that's beyond the desired timeout.  Still a bit fragile (it encodes a lot of logic about the implementation strategy for the timeout - if the implementation changes to use the clock differently, maybe it is still correct but the test starts failing) but maybe it's better than nothing.